cairo/ethereum/cancun/vm/precompiled_contracts/modexp.cairo

```diff
@@ -79,24 +100,308 @@ func modexp{
+        );
+        let (bits_ptr_updated_2, bits_len_updated_2) = extract_limb_bits(
+            num.value.d2, bits_ptr_after_limb1, bits_len_padded
+        );
+        tempvar bits_ptr_after_limb2 = bits_ptr_updated_2;
+        tempvar bits_len_after_limb2 = bits_len_updated_2;
+        tempvar range_check_ptr = range_check_ptr;
+    } else {
+        tempvar bits_ptr_after_limb2 = bits_ptr_after_limb1;
+        tempvar bits_len_after_limb2 = bits_len_after_limb1;
+        tempvar range_check_ptr = range_check_ptr;
+    }
+
+    // Process limb3 (d3)
+    let limb3_not_zero = is_not_zero(num.value.d3);
+    tempvar bits_ptr_after_limb2 = bits_ptr_after_limb2;
+    tempvar range_check_ptr = range_check_ptr;
+    if (limb3_not_zero != 0) {
+        // Use memset to pad with zeros until we reach 288 bits if needed
+        let bits_len_padded = 288;
+        memset(
+            bits_ptr_after_limb2 + bits_len_after_limb2, 0, bits_len_padded - bits_len_after_limb2
+        );
+        let (bits_ptr_updated_3, bits_len_updated_3) = extract_limb_bits(
+            num.value.d3, bits_ptr_after_limb2, bits_len_padded
+        );
+        return (bits_ptr_updated_3, bits_len_updated_3);
+    } else {
+        return (bits_ptr_after_limb2, bits_len_after_limb2);
+    }
+}
+
+func extract_limb_bits{range_check_ptr}(limb: felt, bits_ptr: felt*, current_len: felt) -> (
+    felt*, felt
+) {
+    // Check if limb is zero
+    let is_limb_zero = is_zero(limb);
+    if (is_limb_zero == 1) {
+        return (bits_ptr, current_len);
+    }
+
+    // Extract the least significant bit
+    let (q, r) = divmod(limb, 2);
+    assert bits_ptr[current_len] = r;
+
+    // Continue with the remaining bits
+    return extract_limb_bits(q, bits_ptr, current_len + 1);
+}
+
+// Saturates at 2^128 - 1
+func complexity{range_check_ptr}(base_length: U256, modulus_length: U256) -> Uint {
+    alloc_locals;
+
+    let max_len = U256_max(base_length, modulus_length);
+    tempvar seven_u256 = U256(new U256Struct(7, 0));
+    let (words, overflow) = U256_add_with_carry(max_len, seven_u256);
+    if (overflow != 0) {
+        let result = Uint(2 ** 128 - 1);
+        return result;
+    }
+
+    tempvar eight_u256 = U256(new U256Struct(8, 0));
+    let (quotient, _) = uint256_unsigned_div_rem([words.value], [eight_u256.value]);
+
+    if (quotient.high != 0) {
+        let result = Uint(2 ** 128 - 1);
+        return result;
+    }
+
+    // PRIME - 1
+    tempvar prime_u256 = new Uint256(0, 0x8000000000000011);
+    let (res, carry) = uint256_mul(quotient, quotient);
+    let (res_inf_prime) = uint256_le(res, [prime_u256]);
+    let (carry_is_zero) = uint256_eq(carry, Uint256(0, 0));
+    if (carry_is_zero + res_inf_prime != 2) {
+        let result = Uint(2 ** 128 - 1);
+        return result;
+    }
+
+    let res_felt = uint256_to_felt(res);
+    let res_uint = Uint(res_felt);
+    return res_uint;
```

Comment:
here let's say our computation input is max_length

our output is `words**2`. we need output to saturate at `(2**128 - 1)`. 
which means
```
(words ** 2 <= 2**128 -1) => words 
words <= 2**64 - 1 => (max_length + 7) // 8 <= 2**64 - 1
=> max_length <= 2**67 - 8
```
so you can simplify all these conditions a lot. Either max_length is <= 2**67 - 8 in which case you return the computation, or you saturate directly